Add a few more storage metrics

bertinatto opened the pull request:
**What this PR does / why we need it**:

This PR adds a few metrics described in the [Metrics Spec](https://docs.google.com/document/d/1Fh0T60T_y888LsRwC51CQHO75b2IZ3A34ZQS71s_F0g/edit#heading=h.ys6pjpbasqdu):

Additional metrics for Volume Manager:
* Number of volumes in ActualStateofWorld and DesiredStateofWorld
* Number of times ReconstructVolume Spec on kubelet failed

Additional metrics for A/D Controller:
* Number of volumes in ActualStateofWorld and DesiredStateofWorld
* Numer of times A/D Controller performs force detach

Additional metrics for PV Controller:
* Total provision time
* Total PV deletion time
* Number of times PV provisioning failed
* Number of times PV deletion failed


**Release note**:

```release-note
NONE
```

bertinatto committed:
Add more metrics for A/D Controller

Specifically:

* Number of Volumes in ActualStateofWorld and DesiredStateofWorld
* Numer of times A/D Controller performs force detach

bertinatto committed:
Return error in provisionClaimOperation

bertinatto committed:
Add extra metrics for PV Controller

Specifically:

* Total provision time
* Total PV deletion time
* Number of times PV provisioning failed
* Number of times PV deletion failed

jsafrane commented:
/ok-to-test

bertinatto commented:
/assign @jafrane

bertinatto commented:
/test pull-kubernetes-verify

bertinatto committed:
Add more metrics for Volume Manager

Specifically:

* Number of volumes in ActualStateofWorld and DesiredStateofWorld
* Number of times ReconstructVolume Spec on kubelet failed

bertinatto committed:
Update Bazel files

tsmetana committed:
PV metrics: add failed PV count

jsafrane commented on pkg/kubelet/volumemanager/metrics/metrics.go:
```diff
@@ -20,15 +20,30 @@ import "github.com/prometheus/client_golang/prometheus"
+			Name: "volume_manager_reconstruct_volume_errors_total",
+			Help: "Amount of errors on ReconstructVolumeOperation",
+		},
+		[]string{"volume_name"})
```
You should not use `volume_name` as label, there can be too many of them. From https://prometheus.io/docs/practices/naming/: 

> Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

bertinatto replied:
Gotcha, thanks. I'll remove it.

jsafrane replied:
I could be useful to have plugin name as a label, if the plugin name is easily available.

jsafrane commented on pkg/controller/volume/persistentvolume/metrics/metrics.go:
```diff
@@ -105,6 +106,13 @@ var (
+			Name: "pvc_used_total_count",
+			Help: "Total number of PVCs in use by Pods",
+		},
+		[]string{volumeLabel})
```
Dtto, volume name is not a good label.

jsafrane replied:
Probably no label. However, I don't like where this metric is placed. I don't think it's possible to get total count of PVCs that are used from PV controller without lot of check. @gnufied, is there a better place? Can we collect this metric in kubelet (with node name as label)? One NFS used on several nodes could be counted multiple times in that case, however it might be better metric than nr. of used PVCs - nr. of volumes of each volume plugin for each node.

gnufied replied:
I would like to see this metric with "node" and plugin type as dimension. One catch though is - we will have to use custom collector interface for metrics that are somewhat ephemeral. 

Can we use `pvc_protection` controller as a place where this metric is calculated?

bertinatto replied:
@jsafrane, @gnufied: considering that this PR is already large, would you be OK if I postpone this metric to the next PR?

jsafrane commented on pkg/controller/volume/persistentvolume/metrics/metrics.go:
```diff
@@ -89,8 +91,24 @@ var (
+			Name: "volume_operation_total_seconds",
+			Help: "Total volume operation time",
+		},
+		[]string{"operation_name", storageClassLabel, volumeLabel})
```
Dtto, no label with volume name.

jsafrane reviewed (commented):
What is missing in this PR? Can you remove [WIP]?

bertinatto commented:
> What is missing in this PR? Can you remove [WIP]?

I was planning to add a few more metrics to this PR, namely:

* Numer of times A/D Controller performs force detach
* Volume formatting time
* Number of orphaned pod volume directories
* Number of Attach Failures because of attached to wrong node
* Number of times VerifyVolumesAreAttached mechanism 
fixes volumes

However, I'm OK with postponing this to a later PR if you think it's better.

jsafrane commented:
> I was planning to add a few more metrics to this PR

It's already long enough, I'd rather keep the PRs smaller.

bertinatto commented:
> It's already long enough, I'd rather keep the PRs smaller.

Right, this makes sense. I'll address the pieces you pointed out and then remove the WIP label. Thanks!

bertinatto commented:
/test pull-kubernetes-e2e-gce

bertinatto commented:
/test pull-kubernetes-e2e-gce

bertinatto commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1407,7 +1407,7 @@ func (ctrl *PersistentVolumeController) provisionClaimOperation(claim *v1.Persis
 		glog.V(2).Infof("Mount options are not supported by the provisioner but claim %q's StorageClass %q has mount options %v", claimToClaimKey(claim), storageClass.Name, options.MountOptions)
 		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.ProvisioningFailed, strerr)
-		return
+		return fmt.Errorf("provisioner %q doesn't support mount options", plugin.GetPluginName())
```
@jsafrane, could you verify that this is correct?

jsafrane replied:
Yes, it's correct

bertinatto commented:
/test pull-kubernetes-e2e-gce

jsafrane commented on pkg/kubelet/volumemanager/metrics/metrics.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+// RecordRecVolErrorsMetric registers the amount of errors that happened on ReconstructVolumeOperation.
+func RecordRecVolErrorsMetric(pluginName string) {
```
`RecordReconstructionErrorsMetric`, i.e. "volume" is implied? Or RecordVolumeReconstructionErrorsMetric.

jsafrane commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1322,9 +1327,12 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
-		return nil
+		err := ctrl.provisionClaimOperation(claim)
+		timeTaken := time.Since(startTime).Seconds()
+		metrics.RecordVolOperationMetric("provision", v1helper.GetPersistentVolumeClaimClass(claim), timeTaken, err)
```
It would be great to have the volume plugin as a label.

jsafrane commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1041,8 +1042,12 @@ func (ctrl *PersistentVolumeController) reclaimVolume(volume *v1.PersistentVolum
-			return ctrl.deleteVolumeOperation(volume)
+			err := ctrl.deleteVolumeOperation(volume)
+			timeTaken := time.Since(startTime).Seconds()
+			metrics.RecordVolOperationMetric("delete", volume.Spec.StorageClassName, timeTaken, err)
```
It would be great to have the volume plugin as a label.

bertinatto commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

bertinatto commented:
/test  pull-kubernetes-local-e2e-containerized
/test pull-kubernetes-e2e-gce-device-plugin-gpu

bertinatto commented:
/test pull-kubernetes-local-e2e-containerized

gnufied commented on pkg/kubelet/volumemanager/metrics/metrics.go:
```diff
@@ -0,0 +1,49 @@
+var (
+	volumeManagerVolumesMetric = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "volume_manager_total_volumes",
```
I am not sure if this metric will be super useful without plugin name as a dimension.

gnufied commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -245,6 +247,10 @@ func (rc *reconciler) reconcile() {
+	// Update metrics
+	metrics.RecordADControllerVolumesMetric("actual_state_of_world", len(rc.actualStateOfWorld.GetAttachedVolumes()))
+	metrics.RecordADControllerVolumesMetric("desired_state_of_world", len(rc.desiredStateOfWorld.GetVolumesToAttach()))
+
```
I would remove this from here because there is a certain cost associated with recording a metric and the reconciler loop IMO turns too quickly(100ms) while metric collection will happen every 10s. So prometheus client library will "aggregate" these metrics client side.

So we should consider introducing another timer for collecting these metrics rather than using reconciler.

bertinatto replied:
Done! Thanks for the heads up!

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -154,6 +155,10 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
+		// Update metrics
+		metrics.RecordVolumeManagerVolumesMetric("actual_state_of_world", len(rc.actualStateOfWorld.GetMountedVolumes()))
+		metrics.RecordVolumeManagerVolumesMetric("desired_state_of_world", len(rc.desiredStateOfWorld.GetVolumesToMount()))
 	}
```
Same as before.

gnufied commented on pkg/controller/volume/attachdetach/metrics/metrics.go:
```diff
@@ -0,0 +1,49 @@
+
+// RecordADControllerVolumesMetric registers the amount of volumes, and their respective states, in the A/D Controller.
+func RecordADControllerVolumesMetric(state string, number int) {
+	adControllerVolumesMetric.WithLabelValues(state).Set(float64(number))
```
do we need plugin name here?

gnufied reviewed (commented):
while mostly looks good to me. I think this needs some tests too.

jsafrane commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1041,8 +1042,19 @@ func (ctrl *PersistentVolumeController) reclaimVolume(volume *v1.PersistentVolum
+		startTime := time.Now()
 		ctrl.scheduleOperation(opName, func() error {
-			return ctrl.deleteVolumeOperation(volume)
+			err := ctrl.deleteVolumeOperation(volume)
```
You can return volume plugin name from `deleteVolumeOperation` so you don't need to look it up below.

bertinatto replied:
Done. I also removed the `storage_class` as dimension because it seems that it's not very useful in this case (and to avoid making it more complex).

jsafrane commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1322,16 +1334,22 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
 	ctrl.scheduleOperation(opName, func() error {
-		ctrl.provisionClaimOperation(claim)
-		return nil
+		err := ctrl.provisionClaimOperation(claim)
```
You can return volume plugin name from `provisionClaimOperation` so you don't need to look it up below.

bertinatto commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce

bertinatto commented:
/test pull-kubernetes-integration

bertinatto commented:
/test pull-kubernetes-integration

jsafrane commented:
lgtm-ish, please squash the fixes into original commits

@gnufied, PTAL

bertinatto commented:
/test pull-kubernetes-e2e-gce

gnufied commented:
@bertinatto Can we add some tests please?

bertinatto committed:
Include new metrics in tests

bertinatto commented:
/test pull-kubernetes-e2e-gce-100-performance
/test pull-kubernetes-kubemark-e2e-gce-big

gnufied commented on test/e2e/storage/volume_metrics.go:
```diff
@@ -103,7 +103,15 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
+			"volume_manager_reconstruct_volume_errors_total",
+			"attachdetach_controller_total_volumes",
+			"attachdetach_controller_total_forced_detaches",
+		}
```
Is adding this key enough? How do we ensure that this operation does indeed occur within the test? Have you ran this test?

gnufied commented on test/e2e/storage/volume_metrics.go:
```diff
@@ -170,13 +178,18 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
+			unboundPVCKey           = "pv_collector_unbound_pvc_count"
+			volumeOperationKey      = "volume_operation_total_seconds"
+			volumeOperationErrosKey = "volume_operation_total_errors"
 		)
```
Does the test scenario actually has errors for these metrics to be recorded? We should run these tests in a loop for 20-30mins before we merge, they are known to flake unless carefully verified.

bertinatto replied:
@gnufied, how to you run these tests?

I just noticed that they are skipped [0]. Also, I haven't had success running them locally; one of them always fail, and the other always skip.

[0] https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/63601/pull-kubernetes-e2e-gce/40853/

bertinatto commented:
@gnufied, can you take a look?

gnufied commented on test/e2e/storage/volume_metrics.go:
```diff
@@ -167,6 +167,124 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
+		pod, err = c.CoreV1().Pods(ns).Get(pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		controllerMetrics, err := metricsGrabber.GrabFromControllerManager()
```
We should skip these tests in case controller-manager metrics can not be retrieved.

bertinatto replied:
Done.

gnufied commented on test/e2e/storage/volume_metrics.go:
```diff
@@ -167,6 +167,124 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
+		pod, err = c.CoreV1().Pods(ns).Get(pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		controllerMetrics, err := metricsGrabber.GrabFromControllerManager()
```
Same here. Lets not hard fail the test if controller-manager metrics can not be grabbed.

bertinatto replied:
Done.

gnufied commented on test/e2e/storage/volume_metrics.go:
```diff
@@ -167,6 +167,124 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
+		samples, ok := controllerMetrics[forceDetachKey]
+		Expect(ok).To(BeTrue(), "Key %q not found in A/D Controller metrics", forceDetachKey)
+		for _, sample := range samples {
+			Expect(int64(sample.Value)).To(BeNumerically("==", 0),
```
This could flake in some environments I think. Rather than testing for absolute equality - we should check if this test did not incremenet whatever last value was.

bertinatto replied:
This didn't flake after running for around 30 minutes, but I removed it just in case.

I didn't proceed checking the earlier value because I'm afraid there could be more than one sample, and in that I don't know how if I can make sure that the sample I'm checking against is the one I want.

bertinatto commented:
/retest

bertinatto commented:
/retest

bertinatto commented:
/retest

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-local-e2e-containerized

gnufied commented:
/lgtm

bertinatto commented:
/retest

childsb commented:
/approve

bertinatto commented:
/assign @saad-ali

saad-ali commented on pkg/controller/volume/attachdetach/metrics/metrics.go:
```diff
@@ -134,3 +149,21 @@ func (collector *volumeInUseCollector) getVolumeInUseCount() nodeVolumeCount {
 }
+
+// RecordADControllerVolumesMetric registers the amount of volumes, and their respective states, in the A/D Controller.
+func RecordADControllerVolumesMetric(pluginName, state string) {
```
nit: These methods are called directly, which makes testing the caller difficult because these dependencies can't be easily be mocked out. Unless there is a good reason otherwise, this package should have an interface, and a concrete object that implements that interface.

This file is already like this, so not your fault, but you could clean it up while you're touching the code if you want.

bertinatto replied:
> nit: These methods are called directly, which makes testing the caller difficult because these dependencies can't be easily be mocked out.

From what I've seen around, this seems to be a common pattern with metrics (where tests are usually implemented as e2e). In any case, I refactored this file and left the changes in a separate commit to facilitate the review. Please let me know if I'm on the right track.

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -274,11 +280,33 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 	go adc.desiredStateOfWorldPopulator.Run(stopCh)
+
 	metrics.Register(adc.pvcLister, adc.pvLister, adc.podLister, &adc.volumePluginMgr)
+	go adc.recordADVolumesMetrics(stopCh)
```
Please add unit tests exercising the new code in `attach_detach_controller_test.go`. Hint: this will be hard because of the way this new code is structured. Restructuring the code with interfaces/structs that implement interfaces/dependency injection patterns will allow you test it much more easily and make it more maintainable long term.

saad-ali commented on pkg/controller/volume/attachdetach/metrics/metrics.go:
```diff
@@ -34,6 +34,19 @@ var (
 		"Measure number of volumes in use",
 		[]string{"node", "volume_plugin"}, nil)
+
+	adControllerVolumesMetric = prometheus.NewGaugeVec(
```
A global variable and global methods operating on it is generally not a good idea. It leads to a lot of difficulty unit testing the caller because these dependencies can't be easily replaced. Instead create an interface first defining what methods you need, then create a private struct defining the state you need to implement the interface, then implement a constructor that builds the struct and returns it as an interface parameter, so consumer depends on the interface not the struct implementation.

gnufied replied:
@saad-ali while this is true in other cases - no matter how you slice it, a metric variable is a package global variable and only one instance of it can ever exist in the process. Any attempt to register an already registered metric causes panic errors and process crash. 

From that perspective, i would think that it is better to be explicit and define metrics once in the package and make sure they are initialized/registered only once rather than hide them behind some abstraction, which can potentially cause more errors.  There have been several bugs reported in past which caused controller-manager crash because some code tried to register same metric more than once.

saad-ali reviewed (changes requested):
Started reviewing. Please ping me once these are addressed and I'll take another look. In general, ensure the code is testable.

bertinatto committed:
Refactor AD Controller metrics

bertinatto commented:
/retest

bertinatto commented:
/retest

bertinatto commented:
/retest

bertinatto commented:
Thanks for taking the time to review, @saad-ali! Can you take a look to see if what I implemented is what you meant?

In general, I added e2e tests to make sure that metrics in A/D controller are being collected. The reason is that I found it a bit tricky to do that in unit tests.

bertinatto commented:
@gnufied, since I changed some of the code that you introduced recently, would you mind taking a look as well?

bertinatto commented:
I'll split this PR into smaller ones, so I'm closing this one.
